### ISSUE TITLE
Kinda Huge Security Hole?

### ISSUE THREAD

**Issue by Zcyon** [lenis0012/LoginSecurity-2#203]
**The problem**
I'm pentesting my server since we were the victims of a small troll that managed to get into one of our admin accounts. I saw the logs and figured out how they did it and it turns out it was extremely simple to achieve and if there's no IP comparison (at least) inside the plugin when a session is restored, the fact that we've never been attacked before is beyond me.

**Repro steps**
- Enter your server and log in as usual.
- Open a second client of Minecraft (TLauncher will do since it allows you to set up any name you want) and try to enter that same server with the same username as the first one.
- Notice how your original session is kicked out because "You logged in from another location".
- Notice how your "new session" has been restored "from 0 seconds ago" giving the attacker full access to your account and all they had to do was to wait for you to be logged in (e.g, during a streaming session).

I tried with a `session-timeout` of 0, 1, and -1 and got the same result in all of them.

**Environment info**
- Bukkit server.
- Minecraft version 1.15.12
- LoginSecurity version 3.0.2 (latest? AFAIK) [filename LoginSecurity-3.0.2-Bukkit.jar].

**Important information**
- Again, I STILL have to try with a friend at another location to see if maybe that login attempt (with a different IP) is caught by the plugin, but everything seems to point that it won't be [let's hope for the best, though]. I'll update in the morning but for now, be cautious.

- For more context, I attached my `config` file (as txt). I left `session-timeout` set to -1 AFTER playing around with the settings but it was set to `0` when the attack happened.

[config.txt](https://github.com/lenis0012/LoginSecurity-2/files/5044801/config.txt)

**Comment by Zcyon** [comment]
For more context, this is my `config.yml` file:

`#> LoginSecurity configuration.
#> Some information is provided in the form of comments
#> For more info visit https://github.com/lenis0012/LoginSecurity-2/wiki/Configuration
register:
  required: true
  # When enabled, users need to enter a captcha upon registration.
  captcha: false
  # When enabled, requires users to enter their password twice upon registration.
  confirm-password: false
login:
  max-tries: 5
  # Only allow registered players to join using exactly the same name as registered.
  username-match-exact: true
password:
  min-length: 6
password-max-length: 32
join:
  # When enabled, player gets a blindness effect when not logged in.
  blindness: true
  # Temporarily login location until player has logged in.
  # Available options: DEFAULT, SPAWN, RANDOM
  location: DEFAULT
  # Hides the player's inventory until they log in.
  # The inventory is never lost, even after reboot or crash.
  hide-inventory: true
username:
  # Remove special characters like @ and # from the username.
  # Disabling this can be a security risk!
  filter-special-chars: true
  min-length: 3
  max-length: 16
# Shortcut that can be used as alternative to login/register command. Does not replace the defaults
command-shortcut:
  enabled: false
  login: /l
  register: /reg
updater:
  enabled: true
  # The type of builds you are checking. RELEASE, BETA, ALPHA
  channel: BETA
# Session timeout in seconds, set to -1 to disable.
session-timeout: -1
# Login timeout in seconds, set to -1 to disable.
login-timeout: 120
# Login/register message delay in seconds.
login-message-delay: 10
# Language for messages, check wiki for more info.
# List: https://github.com/lenis0012/Translations
# This setting should be set tot he file name without .json
language: en_us
`

I left `session-timeout` set to -1 after playing around tonight but it was on `0` when the attack happened.

**Comment by Zcyon** [comment]
Came back with my findings:

**tl;dr the plugin is SECURE AF BOIIIIIIII.**

My experiences yesterday were possibly caused because I was accessing from the same IP (which makes total sense). I tried with `session-timeout` set to -1 and 0, and **although the original session was being kicked out, which can still be used to disrupt a stream**, the "attacker" (myself from another location) was stopped by the login screen and I had no access to any OP or admin commands or anything.

However, this means that the cause of the attack my server suffered was a weak password (believe it or not, it's a relief because I was already seeing myself submitting PRs to this plugin).

Closing this as it is no longer relevant.
